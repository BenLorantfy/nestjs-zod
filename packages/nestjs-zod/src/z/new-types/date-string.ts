import {
  INVALID,
  ParseInput,
  ParseReturnType,
  ParseStatus,
  ZodIssueCode,
  ZodParsedType,
  ZodType,
  ZodTypeDef,
} from 'zod'
import {
  addIssueToContext,
  DateStringDayType,
  DateStringDirection,
  DateStringFormat,
} from '../issues'
import {
  ErrorMessage,
  findCheck,
  normalizeErrorMessage,
  processCreateParams,
  RawCreateParams,
} from '../shared'
import { ZodFirstPartyTypeKindExtended } from '../type-names'

type ZodIsoDateCheck =
  | {
      kind: 'format'
      value: DateStringFormat
      regex: RegExp
      message?: string
    }
  | { kind: 'direction'; direction: DateStringDirection; message?: string }
  | { kind: 'day-type'; type: DateStringDayType; message?: string }
  | { kind: 'minYear'; value: number; message?: string }
  | { kind: 'maxYear'; value: number; message?: string }

export interface ZodDateStringDef extends ZodTypeDef {
  checks: ZodIsoDateCheck[]
  typeName: ZodFirstPartyTypeKindExtended.ZodDateString
}

const formatToRegex: Record<DateStringFormat, RegExp> = {
  'date': /^\d{4}-\d{2}-\d{2}$/,
  'date-time':
    /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(|\.\d{3})(Z|[+-]\d{2}:\d{2})$/,
}

export class ZodDateString extends ZodType<string, ZodDateStringDef> {
  _parse(input: ParseInput): ParseReturnType<string> {
    const parsedType = this._getType(input)
    const context = this._getOrReturnCtx(input)

    if (parsedType !== ZodParsedType.string) {
      addIssueToContext(context, {
        code: ZodIssueCode.invalid_type,
        expected: ZodParsedType.string,
        received: context.parsedType,
      })

      return INVALID
    }

    const date = new Date(input.data)

    if (Number.isNaN(date.getTime())) {
      addIssueToContext(context, {
        code: ZodIssueCode.custom,
        // TODO: add message
        message: 'Invalid date string',
        params: {
          isNestJsZod: true,
          code: 'invalid_date_string',
        },
      })

      return INVALID
    }

    const status = new ParseStatus()

    for (const check of this._def.checks) {
      if (check.kind === 'format') {
        const valid = check.regex.test(input.data)
        if (valid) continue

        addIssueToContext(context, {
          code: ZodIssueCode.custom,
          message: check.message,
          params: {
            isNestJsZod: true,
            code: 'invalid_date_string_format',
            expected: check.value,
          },
        })

        status.dirty()
      } else if (check.kind === 'direction') {
        const conditions: Record<DateStringDirection, boolean> = {
          past: date < new Date(),
          future: date > new Date(),
        }

        const valid = conditions[check.direction]
        if (valid) continue

        addIssueToContext(context, {
          code: ZodIssueCode.custom,
          message: check.message,
          params: {
            isNestJsZod: true,
            code: 'invalid_date_string_direction',
            expected: check.direction,
          },
        })

        status.dirty()
      } else if (check.kind === 'day-type') {
        const day = date.getDay()

        const conditions: Record<DateStringDayType, boolean> = {
          weekDay: day !== 0 && day !== 6,
          weekend: day === 0 || day === 6,
        }
        const valid = conditions[check.type]
        if (valid) continue

        addIssueToContext(context, {
          code: ZodIssueCode.custom,
          message: check.message,
          params: {
            isNestJsZod: true,
            code: 'invalid_date_string_day',
            expected: check.type,
          },
        })

        status.dirty()
      } else if (check.kind === 'minYear') {
        const valid = date.getFullYear() >= check.value
        if (valid) continue

        addIssueToContext(context, {
          code: ZodIssueCode.too_small,
          type: 'date_string_year',
          minimum: check.value,
          inclusive: true,
          message: check.message,
        })

        status.dirty()
      } else if (check.kind === 'maxYear') {
        const valid = date.getFullYear() <= check.value
        if (valid) continue

        addIssueToContext(context, {
          code: ZodIssueCode.too_big,
          type: 'date_string_year',
          maximum: check.value,
          inclusive: true,
          message: check.message,
        })

        status.dirty()
      }
    }

    return { status: status.value, value: input.data }
  }

  _replaceCheck(check: ZodIsoDateCheck) {
    return new ZodDateString({
      ...this._def,
      checks: this._def.checks
        .filter((item) => item.kind !== check.kind)
        .concat(check),
    })
  }

  static create = (params?: RawCreateParams): ZodDateString => {
    return new ZodDateString({
      checks: [
        {
          kind: 'format',
          value: 'date-time',
          regex: formatToRegex['date-time'],
        },
      ],
      typeName: ZodFirstPartyTypeKindExtended.ZodDateString,
      ...processCreateParams(params),
    })
  }

  format(format: DateStringFormat, message?: ErrorMessage) {
    return this._replaceCheck({
      kind: 'format',
      value: format,
      regex: formatToRegex[format],
      ...normalizeErrorMessage(message),
    })
  }

  past(message?: ErrorMessage) {
    return this._replaceCheck({
      kind: 'direction',
      direction: 'past',
      ...normalizeErrorMessage(message),
    })
  }

  future(message?: ErrorMessage) {
    return this._replaceCheck({
      kind: 'direction',
      direction: 'future',
      ...normalizeErrorMessage(message),
    })
  }

  weekDay(message?: ErrorMessage) {
    return this._replaceCheck({
      kind: 'day-type',
      type: 'weekDay',
      ...normalizeErrorMessage(message),
    })
  }

  weekend(message?: ErrorMessage) {
    return this._replaceCheck({
      kind: 'day-type',
      type: 'weekend',
      ...normalizeErrorMessage(message),
    })
  }

  minYear(year: number, message?: ErrorMessage) {
    return this._replaceCheck({
      kind: 'minYear',
      value: year,
      ...normalizeErrorMessage(message),
    })
  }

  maxYear(year: number, message?: ErrorMessage) {
    return this._replaceCheck({
      kind: 'maxYear',
      value: year,
      ...normalizeErrorMessage(message),
    })
  }

  cast() {
    return this.transform((string) => new Date(string))
  }

  get format_() {
    return findCheck(this._def.checks, 'format')
  }

  get isPast() {
    return findCheck(this._def.checks, 'direction')?.direction === 'past'
  }

  get isFuture() {
    return findCheck(this._def.checks, 'direction')?.direction === 'future'
  }

  get isWeekDay() {
    return findCheck(this._def.checks, 'day-type')?.type === 'weekDay'
  }

  get isWeekend() {
    return findCheck(this._def.checks, 'day-type')?.type === 'weekend'
  }

  get minYear_() {
    return findCheck(this._def.checks, 'minYear')
  }

  get maxYear_() {
    return findCheck(this._def.checks, 'maxYear')
  }
}

export const dateString = ZodDateString.create
