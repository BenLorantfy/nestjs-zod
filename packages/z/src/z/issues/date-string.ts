import { ZodIssueBase, ZodIssueCode } from 'zod'

export type DateStringFormat = 'date' | 'date-time'
export type DateStringDirection = 'past' | 'future'
export type DateStringDayType = 'weekDay' | 'weekend'

export interface ZodInvalidDateStringIssue extends ZodIssueBase {
  code: typeof ZodIssueCode.custom
  params: {
    isNestJsZod: true
    code: 'invalid_date_string'
  }
}

export interface ZodInvalidDateStringFormatIssue extends ZodIssueBase {
  code: typeof ZodIssueCode.custom
  params: {
    isNestJsZod: true
    code: 'invalid_date_string_format'
    expected: DateStringFormat
  }
}

export interface ZodInvalidDateStringDirectionIssue extends ZodIssueBase {
  code: typeof ZodIssueCode.custom
  params: {
    isNestJsZod: true
    code: 'invalid_date_string_direction'
    expected: DateStringDirection
  }
}

export interface ZodInvalidDateStringDayIssue extends ZodIssueBase {
  code: typeof ZodIssueCode.custom
  params: {
    isNestJsZod: true
    code: 'invalid_date_string_day'
    expected: DateStringDayType
  }
}

export type ZodAnyDateStringIssue =
  | ZodInvalidDateStringIssue
  | ZodInvalidDateStringFormatIssue
  | ZodInvalidDateStringDirectionIssue
  | ZodInvalidDateStringDayIssue
