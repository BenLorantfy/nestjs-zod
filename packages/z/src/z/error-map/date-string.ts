import { ZodIssueCode } from 'zod'
import { DateStringFormat, ZodAnyDateStringIssue } from '../issues'
import { createCustomMapper, createMinMaxMapper } from './shared'

export const dateStringCustom = createCustomMapper<ZodAnyDateStringIssue>(
  (params) => {
    if (params.code === 'invalid_date_string') {
      const message = `Invalid string, expected it to be a valid date`
      return { matched: true, message }
    }

    if (params.code === 'invalid_date_string_format') {
      const mapper: Record<DateStringFormat, string> = {
        'date': 'YYYY-MM-DD (RFC3339 "full-date")',
        'date-time': 'YYYY-MM-DDTHH:mm:ssZ (RFC3339 "date-time")',
      }

      const readable = mapper[params.expected]
      const message = `Invalid date, expected it to match ${readable}`
      return { matched: true, message }
    }

    if (params.code === 'invalid_date_string_direction') {
      const message = `Invalid date, expected it to be the ${params.expected}`
      return { matched: true, message }
    }

    if (params.code === 'invalid_date_string_day') {
      const mapper: Record<typeof params['expected'], string> = {
        weekDay: 'week day',
        weekend: 'weekend',
      }

      const readable = mapper[params.expected]
      const message = `Invalid date, expected it to be a ${readable}`
      return { matched: true, message }
    }

    return { matched: false }
  }
)

export const dateStringYearMinMax = createMinMaxMapper(
  'date_string_year',
  (issue) => {
    if (issue.code === ZodIssueCode.too_small) {
      const appendix = issue.inclusive ? 'or equal to ' : ''
      const message = `Year must be greater than ${appendix}${issue.minimum}`
      return { matched: true, message }
    }

    if (issue.code === ZodIssueCode.too_big) {
      const appendix = issue.inclusive ? 'or equal to ' : ''
      const message = `Year must be less than ${appendix}${issue.maximum}`
      return { matched: true, message }
    }

    return { matched: false }
  }
)
