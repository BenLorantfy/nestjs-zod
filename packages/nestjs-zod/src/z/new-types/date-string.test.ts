import { SafeParseReturnType, ZodSchema } from 'zod'
import { dateString } from './date-string'

describe('dateString', () => {
  let schema: ZodSchema
  let result: SafeParseReturnType<unknown, unknown>

  function is(value: unknown, expected: boolean) {
    result = schema.safeParse(value)
    expect(result.success).toBe(expected)
    if (!result.success) {
      expect(result.error.errors).toMatchSnapshot()
    }
  }

  it('should validate type', () => {
    schema = dateString()
    is(undefined, false)
    is(123, false)
  })

  it('should validate date', () => {
    schema = dateString()
    is('', false)
  })

  it('should validate default format', () => {
    schema = dateString()
    is('2017-08-20', false)
  })

  it('should validate specified format', () => {
    schema = dateString().format('date')
    is('2017-08-20', true)

    schema = dateString().format('date')
    is('2017-08-20T20:00:00Z', false)

    schema = dateString().format('date-time')
    is('2017-08-20T20:00:00Z', true)

    schema = dateString().format('date-time')
    is('2017-08-20', false)
  })

  test('format should override previous format', () => {
    schema = dateString().format('date').format('date-time').format('date')
    is('2017-08-20', true)
  })

  it('should accept all date-time variations', () => {
    schema = dateString().format('date-time')

    is('2017-08-20T20:00:00Z', true)
    is('2017-08-20T20:00:00.000Z', true)
    is('2017-08-20T20:00:00+00:00', true)
    is('2017-08-20T20:00:00-00:00', true)
    is('2017-08-20T20:00:00.000+00:00', true)
    is('2017-08-20T20:00:00.000-00:00', true)
  })

  const past = '2017-08-20T20:00:00Z'
  const future = '2322-08-20T20:00:00Z'

  it('should validate direction', () => {
    schema = dateString().past()

    is(past, true)
    is(future, false)

    schema = dateString().future()

    is(future, true)
    is(past, false)
  })

  test('direction should override previous direction', () => {
    schema = dateString().past().future().past()
    is(past, true)
  })

  const small = '1998-08-20T20:00:00Z'
  const middle = '2022-08-20T20:00:00Z'
  const big = '2322-08-20T20:00:00Z'

  it('should validate min/max year', () => {
    schema = dateString().minYear(2000).maxYear(2100)

    is(small, false)
    is(middle, true)
    is(big, false)
  })

  test('min/max should override previous min/max', () => {
    schema = dateString()
      .minYear(2000)
      .maxYear(2100)
      .minYear(1800)
      .maxYear(2400)

    is(small, true)
    is(middle, true)
    is(big, true)
  })

  it('should validate min and max edge cases', () => {
    schema = dateString().format('date').minYear(2000).maxYear(2100)

    is('2000-08-20', true)
    is('2100-08-20', true)
  })

  it('should validate week days', () => {
    schema = dateString().format('date').weekDay()

    is('2022-05-09', true)
    is('2022-05-10', true)
    is('2022-05-11', true)
    is('2022-05-12', true)
    is('2022-05-13', true)
    is('2022-05-14', false)
    is('2022-05-15', false)

    schema = dateString().format('date').weekend()

    is('2022-05-09', false)
    is('2022-05-10', false)
    is('2022-05-11', false)
    is('2022-05-12', false)
    is('2022-05-13', false)
    is('2022-05-14', true)
    is('2022-05-15', true)
  })

  test('day-type should override previous day-type', () => {
    schema = dateString().format('date').weekDay().weekend()

    is('2022-05-09', false)
    is('2022-05-10', false)
    is('2022-05-11', false)
    is('2022-05-12', false)
    is('2022-05-13', false)
    is('2022-05-14', true)
    is('2022-05-15', true)
  })

  it('should cast to Date when .cast is used', () => {
    schema = dateString().cast()

    const string = '2017-08-20T20:00:00Z'
    is(string, true)
    result = schema.safeParse(string)
    if (!result.success) return
    expect(result.data).toBeInstanceOf(Date)
  })
})
