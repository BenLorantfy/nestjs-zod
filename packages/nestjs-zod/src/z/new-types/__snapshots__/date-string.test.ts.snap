// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dateString day-type should override previous day-type 1`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid input",
    "params": Object {
      "code": "invalid_date_string_day",
      "expected": "weekend",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;

exports[`dateString day-type should override previous day-type 2`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid input",
    "params": Object {
      "code": "invalid_date_string_day",
      "expected": "weekend",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;

exports[`dateString day-type should override previous day-type 3`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid input",
    "params": Object {
      "code": "invalid_date_string_day",
      "expected": "weekend",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;

exports[`dateString day-type should override previous day-type 4`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid input",
    "params": Object {
      "code": "invalid_date_string_day",
      "expected": "weekend",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;

exports[`dateString day-type should override previous day-type 5`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid input",
    "params": Object {
      "code": "invalid_date_string_day",
      "expected": "weekend",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;

exports[`dateString should validate date 1`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid date string",
    "params": Object {
      "code": "invalid_date_string",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;

exports[`dateString should validate default format 1`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid input",
    "params": Object {
      "code": "invalid_date_string_format",
      "expected": "date-time",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;

exports[`dateString should validate direction 1`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid input",
    "params": Object {
      "code": "invalid_date_string_direction",
      "expected": "past",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;

exports[`dateString should validate direction 2`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid input",
    "params": Object {
      "code": "invalid_date_string_direction",
      "expected": "future",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;

exports[`dateString should validate min/max year 1`] = `
Array [
  Object {
    "code": "too_small",
    "inclusive": true,
    "message": "Invalid input",
    "minimum": 2000,
    "path": Array [],
    "type": "date_string_year",
  },
]
`;

exports[`dateString should validate min/max year 2`] = `
Array [
  Object {
    "code": "too_big",
    "inclusive": true,
    "maximum": 2100,
    "message": "Invalid input",
    "path": Array [],
    "type": "date_string_year",
  },
]
`;

exports[`dateString should validate specified format 1`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid input",
    "params": Object {
      "code": "invalid_date_string_format",
      "expected": "date",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;

exports[`dateString should validate specified format 2`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid input",
    "params": Object {
      "code": "invalid_date_string_format",
      "expected": "date-time",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;

exports[`dateString should validate type 1`] = `
Array [
  Object {
    "code": "invalid_type",
    "expected": "string",
    "message": "Required",
    "path": Array [],
    "received": "undefined",
  },
]
`;

exports[`dateString should validate type 2`] = `
Array [
  Object {
    "code": "invalid_type",
    "expected": "string",
    "message": "Expected string, received number",
    "path": Array [],
    "received": "number",
  },
]
`;

exports[`dateString should validate week days 1`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid input",
    "params": Object {
      "code": "invalid_date_string_day",
      "expected": "weekDay",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;

exports[`dateString should validate week days 2`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid input",
    "params": Object {
      "code": "invalid_date_string_day",
      "expected": "weekDay",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;

exports[`dateString should validate week days 3`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid input",
    "params": Object {
      "code": "invalid_date_string_day",
      "expected": "weekend",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;

exports[`dateString should validate week days 4`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid input",
    "params": Object {
      "code": "invalid_date_string_day",
      "expected": "weekend",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;

exports[`dateString should validate week days 5`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid input",
    "params": Object {
      "code": "invalid_date_string_day",
      "expected": "weekend",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;

exports[`dateString should validate week days 6`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid input",
    "params": Object {
      "code": "invalid_date_string_day",
      "expected": "weekend",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;

exports[`dateString should validate week days 7`] = `
Array [
  Object {
    "code": "custom",
    "message": "Invalid input",
    "params": Object {
      "code": "invalid_date_string_day",
      "expected": "weekend",
      "isNestJsZod": true,
    },
    "path": Array [],
  },
]
`;
